### PR TITLE
GEA-1428 dataclasses

### PR DESCRIPTION
**Pull request by pangea-andrest**
- Data classes implementation
- Updates test to work with data classes
- Updates samples apps

### CODE REVIEW

File: pangea/services/audit.py [hunk 1]
```diff
@@ -45,6 +42,305 @@
...
+    action -- auditable action that occurred.
+    new -- The value of a record after it was changed.
+    old -- The value of a record before it was changed.
+    source -- Used to record the location from where an activity occurred.
+    status -- Record whether or not the activity was successful.
+    target -- Used to record the specific record that was targeted by the auditable activity.
+    timestamp -- An optional client-supplied timestamp.
+    """
+
+    message: str | dict
```
Comment (dbenders) [inline pangea/services/audit.py:77]: replace `|` by `Union`

File: pangea/services/audit.py [hunk 2]
```diff
@@ -45,6 +42,305 @@
...
+    event -- Event describing auditable activity.
+    signature -- An optional client-side signature for forgery protection.
+    public_key -- The base64-encoded ed25519 public key used for the signature, if one is provided
+    received_at -- A server-supplied timestamp
+    """
+
+    event: Event
+    signature: Optional[str]
+    public_key: Optional[str]
+    received_at: Optional[str]
```
Comment (dbenders) [inline pangea/services/audit.py:102]: this is not optional, and it's a `datetime`

File: pangea/services/audit.py [hunk 3]
```diff
@@ -45,6 +42,305 @@
...
+
+    message: str | dict
+    actor: Optional[str] = None
+    action: Optional[str] = None
+    new: Optional[str | dict] = None
+    old: Optional[str | dict] = None
+    source: Optional[str] = None
+    status: Optional[str] = None
+    target: Optional[str] = None
+    timestamp: Optional[str] = None
```
Comment (dbenders) [inline pangea/services/audit.py:85]: it's a date time

File: pangea/services/audit.py [hunk 4]
```diff
@@ -45,6 +42,305 @@
 ]
 
 
+class BaseModelConfig(BaseModel):
+    class Config:
+        arbitrary_types_allowed = True
+
+
+class EventVerification(str, enum.Enum):
+    NONE = "none"
```
Comment (dbenders) [inline pangea/services/audit.py:51]: let's think another name, like "couldn't verify"

File: pangea/services/audit.py [hunk 5]
```diff
@@ -45,6 +42,305 @@
...
+    hash -- Event envelope hash.
+    canonical_envelope_base64 -- A base64 encoded canonical JSON form of the event envelope, used for hashing.
+    unpublished_root -- The current unpublished root.
+    membership_proof -- A proof for verifying the unpublished root.
+    consistency_proof -- If prev_root was present in the request, this proof verifies that the new unpublished root is a continuation of the prev_root
+    """
+
+    envelope: Optional[EventEnvelope] = None
+    hash: Optional[str] = None
+    canonical_envelope_base64: Optional[str]
```
Comment (dbenders) [inline pangea/services/audit.py:142]: default is missing (`= None`)

File: pangea/services/audit.py [hunk 6]
```diff
@@ -45,6 +42,305 @@
...
+
+    Arguments:
+    actor -- List of actors to include in search. If empty include all.
+    source -- List of sourcers to include in search. If empty include all.
+    target -- List of targets to include in search. If empty include all.
+    """
+
+    actor: List[str] = []
+    source: List[str] = []
+    target: List[str] = []
```
Comment (dbenders) [inline pangea/services/audit.py:163]: I think there's missing items here, plz check

File: pangea/services/audit.py [hunk 7]
```diff
@@ -45,6 +42,305 @@
...
+    url -- the URL where this root has been published.
+    published_at -- The date/time when this root was published.
+    consistency_proof -- Consistency proof to verify that this root is a continuation of the previous one.
+    """
+
+    tree_name: str
+    size: int
+    root_hash: str
+    url: str
+    published_at: str
```
Comment (dbenders) [inline pangea/services/audit.py:254]: datetime

File: pangea/services/audit.py [hunk 8]
```diff
@@ -45,6 +42,305 @@
...
+    consistency_proof: List[str]
```
Comment (dbenders) [inline pangea/services/audit.py:255]: optional

File: pangea/services/audit.py [hunk 9]
```diff
@@ -45,6 +42,305 @@
...
+
+
+class RootOutput(BaseModelConfig):
+    """
+    Result class after a root request
+    """
+
+    data: Root
```
Comment (dbenders) [inline pangea/services/audit.py:263]: check if we can remove this

File: pangea/services/audit.py [hunk 10]
```diff
@@ -45,6 +42,305 @@
...
+    hash -- The record's hash.
+    leaf_index -- The index of the leaf of the Merkle Tree where this record was inserted.
+    membership_proof -- A cryptographic proof that the record has been persisted in the log.
+    consistency_verification -- Consistency verification calculated if required.
+    membership_verification -- Membership verification calculated if required.
+    signature_verification -- Signature verification calculated if required.
+    """
+
+    envelope: EventEnvelope
+    hash: str
```
Comment (dbenders) [inline pangea/services/audit.py:281]: optional

File: pangea/services/audit.py [hunk 11]
```diff
@@ -45,6 +42,305 @@
...
+    expires_at -- The time when the results will no longer be available to page through via the results API.
+    count -- The total number of results that were returned by the search.
+    root -- A root of a Merkle Tree.
+    events -- A list of matching audit records.
+    """
+
+    count: int
+    events: List[SearchEvent]
+    id: Optional[str] = None
+    expires_at: Optional[str] = None
```
Comment (dbenders) [inline pangea/services/audit.py:304]: datetime

File: pangea/services/audit.py [hunk 12]
```diff
@@ -45,6 +42,305 @@
...
+
+    Arguments:
+    count -- The total number of results that were returned by the search.
+    events -- A list of matching audit records.
+    root -- A root of a Merkle Tree.
+    """
+
+    count: int
+    events: List[SearchEvent]
+    root: Root
```
Comment (dbenders) [inline pangea/services/audit.py:341]: optional

File: pangea/response.py [hunk 13]
```diff
@@ -28,53 +30,84 @@ def __setattr__(self, name: str, value) -> None:
...
-    Properties:
-        result (obj): "result" field of the API response as documented at:
-            [https://docs.dev.pangea.cloud/docs/api/#responses](https://docs.dev.pangea.cloud/docs/api/#responses)
-        status (str): Pangea status code
-        status_code(int): HTTP Status Code
-        success (bool): true if call was successful
-        request_id (str): the ID of the request as tracked by Pangea
-        response (obj): the entire API response payload
 
+T = TypeVar("T")
```
Comment (dbenders) [inline pangea/response.py:38]: are you finally using this T?

File: pangea/services/audit.py [hunk 14]
```diff
@@ -45,6 +42,305 @@
...
+    """
+
+    query: str
+    order: Optional[SearchOrder] = None
+    order_by: Optional[SearchOrderBy] = None
+    last: Optional[str] = None
+    start: Optional[datetime.time] = None
+    end: Optional[datetime.time] = None
+    limit: Optional[int] = None
+    max_results: Optional[int] = None
```
Comment (dbenders) [inline pangea/services/audit.py:218]: not now, but maybe you can use pedantic to validate the inputs here (for example, max_results should be lower than 10000 I think)

File: pangea/services/audit.py [hunk 15]
```diff
@@ -45,6 +42,305 @@
...
+class RootInput(BaseModelConfig):
+    """
+    Input class to perform a root request action
+
+    Arguments:
+
+    tree_size -- The size of the tree (the number of records).
+    """
+
+    tree_size: int
```
Comment (dbenders) [inline pangea/services/audit.py:234]: Optional!

File: pangea/services/audit.py [hunk 16]
```diff
@@ -430,74 +677,55 @@ def results(self, id: str, limit: int = 20, offset: int = 0, verify_signatures:
 
         endpoint_name = "results"
 
-        if not id:
-            raise AuditException("An 'id' parameter is required")
-
-        if not (isinstance(limit, int) and limit > 0):
+        if input.limit <= 0:
```
Comment (dbenders) [inline pangea/services/audit.py:680]: this validation can be done in the data class with pedantic

File: pangea/services/embargo.py [hunk 17]
```diff
@@ -1,10 +1,68 @@
...
+class Sanction(BaseModelConfig):
+    """
+    TODO: complete
+    """
+
+    embargoed_country_iso_code: str
+    issuing_country: str
+    list_name: str
+    embargoed_country_name: str
+    annotations: dict[str, Any]
```
Comment (dbenders) [inline pangea/services/embargo.py:52]: Any implies str

File: pangea/services/redact.py [hunk 18]
```diff
@@ -2,15 +2,112 @@
...
+    data_ket --
+    """
+
+    field_type: str
+    score: int
+    text: str
+    start: int
+    end: int
+    redacted: bool
+    data_key: Optional[str]
```
Comment (dbenders) [inline pangea/services/redact.py:58]: = None

File: pangea/services/redact.py [hunk 19]
```diff
@@ -2,15 +2,112 @@
...
+    Class input to redact structured data request
+
+    Arguments:
+    data -- Structured data to redact
+    jsonp -- JSON path(s) used to identify the specific JSON fields to redact in the structured data. Note: If jsonp parameter is used, the data parameter must be in JSON format.
+    format -- The format of the structured data to redact. (default is JSON)
+    debug -- Setting this value to true will provide a detailed analysis of the redacted data and the rules that caused redaction.
+    """
+
+    data: dict | str
```
Comment (dbenders) [inline pangea/services/redact.py:95]: Union

File: pangea/services/redact.py [hunk 20]
```diff
@@ -2,15 +2,112 @@
...
+    debug: Optional[bool] = None
+
+
+class StructuredOutput(PangeaResponseResult):
+    """
+    TODO: complete
+
+    """
+
+    redacted_data: str | Dict  # FIXME: this should be raw json
```
Comment (dbenders) [inline pangea/services/redact.py:107]: Union

File: pangea/signing.py [hunk 21]
```diff
@@ -55,7 +57,7 @@ def _signMessageJSON(self, messageJSON: dict, private_key: ed25519.Ed25519Privat
         message_bytes = canonicalize_json(messageJSON)
         return self._signMessageBytes(message_bytes, private_key)
 
-    def signMessage(self, message: t.Any):
+    def signMessage(self, message: str | dict | bytes):
```
Comment (dbenders) [inline pangea/signing.py:60]: Union

File: pangea/signing.py [hunk 22]
```diff
@@ -79,9 +81,8 @@ def getPublicKeyBytes(self):
 
 class Verifier:
     # verify message with signature and public key bytes
-    def verifyMessage(
-        self, signature_b64: bytes, message: t.Union[str, dict, bytes], public_key_bytes: bytes = None
-    ) -> bool:
+    def verifyMessage(self, signature_b64: str, message: t.Union[str, dict, bytes], public_key_b64: str = None) -> bool:
```
Comment (dbenders) [inline pangea/signing.py:84]: I don't quite like this polymorphic function, I prefer to call the correct one (not for changing now)

File: examples/audit/audit_complete.py [hunk 23]
```diff
@@ -1,107 +1,87 @@
 import os
 
 from pangea.config import PangeaConfig
+from pangea.exceptions import AuditException
```
Comment (dbenders) [inline examples/audit/audit_complete.py:4]: why do we have two files here, `audit_complete.py` and `audit.py`

File: pangea/response.py (hunk 13 above)
Comment (pangea-andrest) [reply pangea/response.py:38]: In some way, yes.

File: pangea/response.py [hunk 24]
```diff
@@ -28,53 +30,84 @@ def __setattr__(self, name: str, value) -> None:
...
+    pass
 
-    @property
-    def request_id(self) -> t.Optional[str]:
-        return self._data.get("request_id", None)
 
-    @property
-    def response(self) -> requests.Response:
-        return self._response
+class ResponseStatus(str, enum.Enum):
```
[the remark this answers is not in the document]
Comment (pangea-andrest) [reply pangea/response.py:72]: Yes. Fixed.

File: pangea/response.py [hunk 25]
```diff
@@ -28,53 +30,84 @@ def __setattr__(self, name: str, value) -> None:
...
+    request_time -- The time the request was issued, ISO8601.
+    response_time -- The time the response was issued, ISO8601.
+    status -- The HTTP status code msg.
+    summary -- The summary of the response.
+    """
+
+    request_id: str
+    request_time: str
+    response_time: str
+    status: str
```
[the remark this answers is not in the document]
Comment (pangea-andrest) [reply pangea/response.py:92]: Fixed

File: pangea/signing.py (hunk 22 above)
Comment (pangea-andrest) [reply pangea/signing.py:84]: In fact, I think it's always call with bytes, but I'll check it in another PR.

File: pangea/services/embargo.py (hunk 17 above)
Comment (pangea-andrest) [reply pangea/services/embargo.py:52]: I'm not sure if I understand. Sometimes I think it's another dict

File: examples/audit/audit_complete.py (hunk 23 above)
Comment (dbenders) [reply examples/audit/audit_complete.py:4]: I think we should have one :)

**Pull request merged**